pydantic/_internal/_generate_schema.py

```diff
@@ -783,25 +776,7 @@ def _generate_schema_from_property(self, obj: Any, source: Any) -> core_schema.C
                     '`__get_validators__` is deprecated and will be removed, use `__get_pydantic_core_schema__` instead.',
                     PydanticDeprecatedSince20,
                 )
-            schema = core_schema.chain_schema([core_schema.with_info_plain_validator_function(v) for v in validators()])
-        else:
-            # we have no existing schema information on the property, exit early so that we can go generate a schema
-            return None
-
-        schema = self._unpack_refs_defs(schema)
-
-        if is_function_with_inner_schema(schema):
-            ref = schema['schema'].pop('ref', None)  # pyright: ignore[reportCallIssue, reportArgumentType]
-            if ref:
-                schema['ref'] = ref
-        else:
-            ref = get_ref(schema)
```

Comment:
This is related to what I mentioned:

> the first one I could find: https://github.com/pydantic/pydantic/issues/7102. The https://github.com/pydantic/pydantic/issues/7102#issuecomment-1682288722 isn't really compelling, the user was doing things not the intended way. It also led to [a scary and hacky fix](https://github.com/pydantic/pydantic/pull/8066/files) that I would really like not having.

These removed lines had the effect of moving the ref from the inner schema of a `function-*` schema to the the `function-*` schema itself. With the following simplified test code added in the mentioned issue above:

```python
class Numeric(BaseModel):
    value: float

    @classmethod
    def __get_pydantic_core_schema__(cls, source_type, handler):
        return core_schema.no_info_before_validator_function(cls.validate, handler(source_type))

    @classmethod
    def validate(cls, v):
        ...

class OuterModel(BaseModel):
    x: Numeric
    y: Numeric
```

On `main`, the schema of `OuterModel` would look like:

<details>

```python
{
│   'type': 'definitions',
│   'schema': {
│   │   'type': 'model',
│   │   'cls': <class '__main__.OuterModel'>,
│   │   'schema': {
│   │   │   'type': 'model-fields',
│   │   │   'fields': {
│   │   │   │   'x': {'type': 'model-field', 'schema': {'type': 'definition-ref', 'schema_ref': '__main__.Numeric:109238506193520'}, 'metadata': {}},
│   │   │   │   'y': {'type': 'model-field', 'schema': {'type': 'definition-ref', 'schema_ref': '__main__.Numeric:109238506193520'}, 'metadata': {}}
│   │   │   },
│   │   │   'model_name': 'OuterModel',
│   │   │   'computed_fields': []
│   │   },
│   │   'config': {'title': 'OuterModel'},
│   │   'ref': '__main__.OuterModel:109238503123056',
│   │   'metadata': {'<stripped>'}
│   },
│   'definitions': [
│   │   {
│   │   │   'function': {'type': 'no-info', 'function': <bound method Numeric.validate of <class '__main__.Numeric'>>},
│   │   │   'schema': {
│   │   │   │   'type': 'model',
│   │   │   │   'cls': <class '__main__.Numeric'>,
│   │   │   │   'schema': {
│   │   │   │   │   'type': 'model-fields',
│   │   │   │   │   'fields': {'value': {'type': 'model-field', 'schema': {'type': 'float'}, 'metadata': {}}},
│   │   │   │   │   'model_name': 'Numeric',
│   │   │   │   │   'computed_fields': []
│   │   │   │   },
│   │   │   │   'config': {'title': 'Numeric'}
│   │   │   },
│   │   │   'ref': '__main__.Numeric:109238506193520',
│   │   │   'metadata': {'<stripped>'},
│   │   │   'type': 'function-before'
│   │   }
│   ]
}
```

</details>

On this PR, it looks like:

<details>

```python
{
│   'type': 'definitions',
│   'schema': {
│   │   'type': 'model',
│   │   'cls': <class '__main__.OuterModel'>,
│   │   'schema': {
│   │   │   'type': 'model-fields',
│   │   │   'fields': {
│   │   │   │   'x': {
│   │   │   │   │   'type': 'model-field',
│   │   │   │   │   'schema': {
│   │   │   │   │   │   'function': {'type': 'no-info', 'function': <bound method Numeric.validate of <class '__main__.Numeric'>>},
│   │   │   │   │   │   'schema': {
│   │   │   │   │   │   │   'function': {'type': 'no-info', 'function': <bound method Numeric.validate of <class '__main__.Numeric'>>},
│   │   │   │   │   │   │   'schema': {'type': 'definition-ref', 'schema_ref': '__main__.Numeric:105945921898336'},
│   │   │   │   │   │   │   'metadata': {'<stripped>'},
│   │   │   │   │   │   │   'type': 'function-before'
│   │   │   │   │   │   },
│   │   │   │   │   │   'metadata': {'<stripped>'},
│   │   │   │   │   │   'type': 'function-before'
│   │   │   │   │   },
│   │   │   │   │   'metadata': {}
│   │   │   │   },
│   │   │   │   'y': {
│   │   │   │   │   'type': 'model-field',
│   │   │   │   │   'schema': {
│   │   │   │   │   │   'function': {'type': 'no-info', 'function': <bound method Numeric.validate of <class '__main__.Numeric'>>},
│   │   │   │   │   │   'schema': {
│   │   │   │   │   │   │   'function': {'type': 'no-info', 'function': <bound method Numeric.validate of <class '__main__.Numeric'>>},
│   │   │   │   │   │   │   'schema': {'type': 'definition-ref', 'schema_ref': '__main__.Numeric:105945921898336'},
│   │   │   │   │   │   │   'metadata': {'<stripped>'},
│   │   │   │   │   │   │   'type': 'function-before'
│   │   │   │   │   │   },
│   │   │   │   │   │   'metadata': {'<stripped>'},
│   │   │   │   │   │   'type': 'function-before'
│   │   │   │   │   },
│   │   │   │   │   'metadata': {}
│   │   │   │   }
│   │   │   },
│   │   │   'model_name': 'OuterModel',
│   │   │   'computed_fields': []
│   │   },
│   │   'config': {'title': 'OuterModel'},
│   │   'ref': '__main__.OuterModel:105945922827312',
│   │   'metadata': {'<stripped>'}
│   },
│   'definitions': [
│   │   {
│   │   │   'type': 'model',
│   │   │   'cls': <class '__main__.Numeric'>,
│   │   │   'schema': {
│   │   │   │   'type': 'model-fields',
│   │   │   │   'fields': {'value': {'type': 'model-field', 'schema': {'type': 'float'}, 'metadata': {}}},
│   │   │   │   'model_name': 'Numeric',
│   │   │   │   'computed_fields': []
│   │   │   },
│   │   │   'config': {'title': 'Numeric'},
│   │   │   'ref': '__main__.Numeric:105945921898336'
│   │   }
│   ]
}
```

</details>

Essentially, the difference in these two schemas is that we don't "move" the ref from the inner schema to the `function-before` schemas.

The changes in this PR + removing this reference moving coincidentally make it work still.

However, doing so was a dangerous game: on _L793_, `schema` directly comes from another model. The `pop` calls removes the reference to the schema, and mutating schemas from other models has been a known issue. 

You may be wondering: why this doesn't break things in the example I gave? Surely the `pop` call should have mutated the core schema of `Numeric`. Turns out it doesn't, because `Numeric.__get_pydantic_core_schema__` does not cache the schema, so calling it will generate a new one every time (and this is what happens during the schema gen of `OuterModel`). But on a similar issue, [I mentioned](https://github.com/pydantic/pydantic/issues/10160#issuecomment-2298257506) that explicitly caching the core schema in the `__get_pydantic_core_schema__` method would resolve the user issue (as the use case was slightly different)! 

So to conclude, overriding `BaseModel.__get_pydantic_core_schema__` is full of unexpected behaviors, but that's fine as officially supporting them would be a huge pain.